README.md
# <img src="https://github.com/pip-services/pip-services/raw/master/design/Logo.png" alt="Pip.Services Logo" style="max-width:30%"> <br/> Mappings microservice

This is Mappings microservice from Pip.Services library. 
It stores Mappings between internally and external service

The microservice currently supports the following deployment options:
* Deployment platforms: Standalone Process, Seneca
* External APIs: HTTP/REST
* Persistence: Flat Files, MongoDB

This microservice has no dependencies on other microservices.

<a name="links"></a> Quick Links:

* [Download Links](doc/Downloads.md)
* [Development Guide](doc/Development.md)
* [Configuration Guide](doc/Configuration.md)
* [Deployment Guide](doc/Deployment.md)
* Mappings Service
  - [Node.js SDK](https://github.com/pip-services/pip-services-mappings-node)
* Communication Protocols
  - [HTTP Version 1](doc/HttpProtocolV1.md)
 
## Contract

Logical contract of the microservice is presented below. For physical implementation (HTTP/REST, Thrift, Seneca, Lambda, etc.),
please, refer to documentation of the specific protocol.

```typescript
class MappingV1 implements IStringIdentifiable {
    
    public id:string;
    public collection: string;
    public internal_id: string;
    public external_id: string;
    public expiration_time: Date;

}

interface IMappingsController {
    getCollectionNames(correlationId: string, callback: (err: any, items: Array<string>) => void);
    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams, callback: (err: any, page: DataPage<MappingV1>) => void);
    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number, callback: (err: any) => void);
    mapToExternal(correlationId: string, collection: string, internalId: string, callback: (err: any, externalId: string) => void);
    mapToInternal(correlationId: string, collection: string, externalId: string, callback: (err: any, internalId: string) => void);
    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string, callback: (err: any) => void);
    deleteExpiredMappings(correlationId: string, callback: (err: any) => void);
}
```

## Download

Right now the only way to get the microservice is to check it out directly from github repository
```bash
git clone git@github.com:pip-services-integration/pip-clients-mappings-node.git
```

Pip.Service team is working to implement packaging and make stable releases available for your 
as zip downloadable archieves.

## Run

Add **config.yml** file to the root of the microservice folder and set configuration parameters.
As the starting point you can use example configuration from **config.example.yml** file. 

Example of microservice configuration
```yaml
- descriptor: "pip-services-container:container-info:default:default:1.0"
  name: "pip-services-mappings"
  description: "Mappings microservice"

- descriptor: "pip-services-commons:logger:console:default:1.0"
  level: "trace"

- descriptor: "pip-services-mappings:persistence:file:default:1.0"
  path: "./data/mappings.json"

- descriptor: "pip-services-mappings:controller:default:default:1.0"

- descriptor: "pip-services-mappings:service:http:default:1.0"
  connection:
    protocol: "http"
    host: "0.0.0.0"
    port: 8080
```
 
For more information on the microservice configuration see [Configuration Guide](Configuration.md).

Start the microservice using the command:
```bash
node run
```

## Install

The easiest way to work with the microservice is to use client SDK. 
The complete list of available client SDKs for different languages is listed in the [Quick Links](#links)

If you use Node.js then you should add dependency to the client SDK into **package.json** file of your project
```javascript
{
    ...
    "dependencies": {
        ....
        "pip-clients-mappings-node": "^1.0.*",
        ...
    }
}
```


## Use

Inside your code get the reference to the client SDK
```typescript
 import { MappingsHttpClientV1 } from 'pip-clients-mappings-node';
```

Define client configuration parameters.

```typescript
// Client configuration
var httpConfig = ConfigParams.fromTuples(
    "connection.protocol", "http",
    "connection.host", "localhost",
    "connection.port", 3000
);
client.configure(httpConfig);
```

Instantiate the client and open connection to the microservice
```typescript
// Create the client instance
client = new MappingsHttpClientV1();

// Connect to the microservice
client.open(null, function(err) {
    if (err) {
        console.error('Connection to the microservice failed');
        console.error(err);
        return;
    }
    
    // Work with the microservice
    ...
});
```
Now the client is ready to perform operations:

Add mapping:
```typescript 
client.addMapping("123", "Common.Collection", "123", "789", 60 * 1000, (err, mapping) => {
  if (err != null) {
    console.error('Can\'t create mapping!');
    console.error(err);
  } else {
    console.dir('Mapping was created successfull');
  }
});
```

Get mappings by filter:
```typescript    

    client.getMappings("123", new FilterParams(), new PagingParams(), (err, page) => {
        if (err != null) {
            console.error('Can\'t get mappings!');
            console.error(err);
        } else {
            console.dir('Mappings was recived successfull');
            for (let mapping in page.data) {
                console.dir('Mappings:');
                console.dir(mapping.toString());
            }
        }
    });

    client.getMappings(null, FilterParams.fromTuples("collection", "Common.Collection"), 
      new PagingParams(1, 10, false),  (err, page) => {
        if (err != null) {
            console.error('Can\'t get mappings!');
            console.error(err);
        } else {
            console.dir('Mappings was recived successfull');
            for (let mapping in page.data) {
                console.dir('Mappings:');
                console.dir(mapping.toString());
            }
        }
    });

```

Delete mapping:
```typescript    

    client.deleteMappingById("123", "Common.Collection", "123", "789", (err) => {
        if (err != null) {
            console.error('Can\'t delete mapping!');
            console.error(err);
        } else {
            console.dir('Mapping was delete successfull');
        }
    });
```

## Control functions

Get collection names:
```typescript
    client.getCollectionNames("123", (err, collections) => {
        if (err != null) {
            bconsole.error('Can\'t get collection names');
            console.error(err);
        } else {
            console.dir('Collections names:');
            console.dir(collections);
        }
    });
```

Get external id by collection name and internal id:
```typescript
    client.mapToExternal("123", "Common.Collection", "123", (err, id) => {
        if (err != null) {
            console.error('Can\'t get external id!');
            console.error(err);
        } else {
            console.dir('External id:', id);
        }
    });
```

Get internal id by collection name and external id:
```typescript
    client.mapToInternal("123", "Common.Collection", "789", (err, id)) => {
        if (err != null) {
            console.error('Can\'t compleate mapping!');
            console.error(err);
        } else {
            console.dir('Internal id:', id);
            }
    }); 
```


## Acknowledgements

This microservice was created and currently maintained by 
- *Sergey Seroukhov*.
- *Levichev Dmitry*


CHANGELOG.md
# Mappings Client SDK ChangeLog

## <a name="1.0.0"></a> 1.0.0 (2020-07-10)

### Features
* **MongoDB** data access using Mongoose framework
* **File** data access 
* **Memory** data access 
* **REST** API client and service
* **Unit tests** using Mocha and Matcha

### Breaking Changes
No breaking changes since this is the first version

### Bug Fixes
No fixes in this version



test/version1/MappingsClientV1Fixture.ts
let _ = require('lodash');
let async = require('async');
let assert = require('chai').assert;

import { FilterParams } from 'pip-services3-commons-node';
import { PagingParams } from 'pip-services3-commons-node';

import { IMappingsClient } from '../../src/version1/IMappingsClient';

export class MappingsClientV1Fixture {
    private _client: IMappingsClient;

    constructor(persistence) {
        assert.isNotNull(persistence);
        this._client = persistence;
    }

    testGetMappingCollections(done) {
        async.series([
            // Add mappings
            (callback) => {
                this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000, callback);
            }, (callback) => {
                this._client.getCollectionNames(null, (err, items) => {
                    assert.isNull(err);
                    assert.equal(2, items.length);
                    assert.include(items, "Common.Collection");
                    assert.include(items, "Common.AnotherCollection");
                    callback();
                });
            }
        ], done);
    }

    testGetMappings(done) {
        async.series([// Add mappings
            (callback) => {
                this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.Collection", "AAA", "111", 60 * 1000, callback);
            }, (callback) => {
                this._client.getMappings(null, FilterParams.fromTuples("collection", "Common.Collection"), new PagingParams(1, 10, false), (err, mappings) => {
                    assert.isNull(err);
                    assert.isNotNull(mappings.data);
                    assert.equal(2, mappings.data.length);
                    callback();
                });

            }
        ], done);
    }

    testMapping(done) {
        async.series([
            // Add mappings
            (callback) => {
                this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000, callback);
            }, (callback) => {
                this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000, callback);
            }, (callback) => {

                // Test internal mappings
                this._client.mapToExternal(null, "Common.Collection", "123", (err, id) => {
                    assert.isNull(err);
                    assert.equal("789", id);
                    callback();
                });

            }, (callback) => {
                // Test external mappings
                this._client.mapToInternal(null, "Common.Collection", "789", (err, id) => {
                    assert.isNull(err);
                    assert.equal("123", id);
                    callback();
                });

            }, (callback) => {
                // Test different collection
                this._client.mapToExternal(null, "Common.AnotherCollection", "123", (err, id) => {
                    assert.isNull(err);
                    assert.equal("543", id);
                    callback();
                });

            }, (callback) => {
                // Test non-exiting collection
                this._client.mapToExternal(null, "Common.YetAnotherCollection", "123", (err, id) => {
                    assert.isNull(err);
                    assert.isNull(id);
                    callback();
                });

            }, (callback) => {
                // Test non-exiting mapping
                this._client.mapToExternal(null, "Common.Collection", "555", (err, id) => {
                    assert.isNull(err);
                    assert.isNull(id);
                    callback();
                });

            }
        ], done);
    }

}


test/version1/MappingsHttpClient.test.ts
import { ConfigParams } from 'pip-services3-commons-node';
import { Descriptor } from 'pip-services3-commons-node';
import { References } from 'pip-services3-commons-node';

import { MappingsMemoryPersistence } from 'pip-services-mappings-node';
import { MappingsController } from 'pip-services-mappings-node';
import { MappingsHttpServiceV1 } from 'pip-services-mappings-node';

import { MappingsHttpClientV1 } from '../../src/version1/MappingsHttpClientV1';
import { MappingsClientV1Fixture } from './MappingsClientV1Fixture';

suite('MappingsHttpClientV1', () => {
    let persistence: MappingsMemoryPersistence;
    let controller: MappingsController;
    let service: MappingsHttpServiceV1;
    let client: MappingsHttpClientV1;
    let fixture: MappingsClientV1Fixture;

    setup((done) => {
        persistence = new MappingsMemoryPersistence();
        persistence.configure(new ConfigParams());

        controller = new MappingsController();
        controller.configure(new ConfigParams());

        let httpConfig = ConfigParams.fromTuples(
            'connection.protocol', 'http',
            'connection.port', 3000,
            'connection.host', 'localhost'
        );

        service = new MappingsHttpServiceV1();
        service.configure(httpConfig);

        client = new MappingsHttpClientV1();
        client.configure(httpConfig);

        let references = References.fromTuples(
            new Descriptor('pip-services-mappings', 'persistence', 'memory', 'default', '1.0'), persistence,
            new Descriptor('pip-services-mappings', 'controller', 'default', 'default', '1.0'), controller,
            new Descriptor('pip-services-mappings', 'service', 'http', 'default', '1.0'), service,
            new Descriptor('pip-services-mappings', 'client', 'http', 'default', '1.0'), client
        );
        controller.setReferences(references);
        service.setReferences(references);
        client.setReferences(references);

        fixture = new MappingsClientV1Fixture(client);

        persistence.open(null, (err) => {
            if (err) {
                done(err);
                return;
            }

            service.open(null, (err) => {
                if (err) {
                    done(err);
                    return;
                }

                client.open(null, done);
            });
        });
    });

    teardown((done) => {
        client.close(null, (err) => {
            service.close(null, (err) => {
                persistence.close(null, done);
            });
        });
    });

    test('Get Mapping Collections', (done) => {
        fixture.testGetMappingCollections(done);
    });

    test('Get Mappings', (done) => {
        fixture.testGetMappings(done);
    });

    test('Mapping', (done) => {
        fixture.testMapping(done);
    });

});

test/version1/MappingsDirectClientV1.test.ts

import { ConfigParams } from 'pip-services3-commons-node';
import { Descriptor } from 'pip-services3-commons-node';
import { References } from 'pip-services3-commons-node';

import { MappingsMemoryPersistence } from 'pip-services-mappings-node';
import { MappingsController } from 'pip-services-mappings-node';

import { MappingsDirectClientV1 } from '../../src/version1/MappingsDirectClientV1';
import { MappingsClientV1Fixture } from './MappingsClientV1Fixture';


suite('MappingsDirectClientV1', () => {
    let persistence: MappingsMemoryPersistence;
    let controller: MappingsController;
    let client: MappingsDirectClientV1;
    let fixture: MappingsClientV1Fixture;

    setup((done) => {
        persistence = new MappingsMemoryPersistence();
        persistence.configure(new ConfigParams());

        controller = new MappingsController();
        controller.configure(new ConfigParams());

        client = new MappingsDirectClientV1();

        let references = References.fromTuples(
            new Descriptor('pip-services-mappings', 'persistence', 'memory', 'default', '1.0'), persistence,
            new Descriptor('pip-services-mappings', 'controller', 'default', 'default', '1.0'), controller,
            new Descriptor('pip-services-mappings', 'client', 'direct', 'default', '1.0'), client
        );

        controller.setReferences(references);
        client.setReferences(references);

        fixture = new MappingsClientV1Fixture(client);

        persistence.open(null, done);
    });

    teardown((done) => {
        persistence.close(null, done);
    });

    test('Get Mapping Collections', (done) => {
        fixture.testGetMappingCollections(done);
    });

    test('Get Mappings', (done) => {
        fixture.testGetMappings(done);
    });

    test('Mapping', (done) => {
        fixture.testMapping(done);
    });

});

src/version1/IMappingsClient.ts
import { DataPage } from 'pip-services3-commons-node';
import { FilterParams } from 'pip-services3-commons-node';
import { PagingParams } from 'pip-services3-commons-node';

import { MappingV1 } from './MappingV1';

export interface IMappingsClient {
    getCollectionNames(correlationId: string,
        callback: (err: any, items: Array<string>) => void);

    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams,
        callback: (err: any, page: DataPage<MappingV1>) => void);

    addMapping(correlationId: string, collection: string,
        internalId: string, externalId: string, timeToLive: number,
        callback: (err: any) => void);

    mapToExternal(correlationId: string, collection: string, internalId: string,
        callback: (err: any, externalId: string) => void);

    mapToInternal(correlationId: string, collection: string, externalId: string,
        callback: (err: any, internalId: string) => void);

    deleteMapping(correlationId: string, collection: string,
        internalId: string, externalId: string,
        callback: (err: any) => void);
}


src/version1/MappingsHttpClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams } from "pip-services3-commons-node";
import { CommandableHttpClient } from 'pip-services3-rpc-node';

export class MappingsHttpClientV1 extends CommandableHttpClient implements IMappingsClient {

    public constructor() {
        super('v1/mappings');
    }

    getCollectionNames(correlationId: string, callback: (err: any, items: string[]) => void) {
        this.callCommand(
            'get_collection_names',
            correlationId,
            {},
            callback
        );
    }

    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams, callback: (err: any, page: import("pip-services3-commons-node").DataPage<import("./MappingV1").MappingV1>) => void) {
        this.callCommand(
            'get_mappings',
            correlationId,
            {
                filter: filter,
                paging: paging
            },
            callback
        );
    }

    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number, callback: (err: any) => void) {
        this.callCommand(
            'add_mapping',
            correlationId,
            {
                collection: collection,
                internal_id: internalId,
                external_id: externalId,
                ttl: timeToLive
            }, (err, res) => {
                callback(err);
            }
        );
    }

    mapToExternal(correlationId: string, collection: string, internalId: string, callback: (err: any, externalId: string) => void) {
        this.callCommand(
            'map_to_external',
            correlationId,
            {
                collection: collection,
                internal_id: internalId
            },
            callback
        );
    }

    mapToInternal(correlationId: string, collection: string, externalId: string, callback: (err: any, internalId: string) => void) {
        this.callCommand(
            'map_to_internal',
            correlationId,
            {
                collection: collection,
                external_id: externalId
            },
            callback
        );
    }

    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string, callback: (err: any) => void) {
        this.callCommand(
            'delete_mapping',
            correlationId,
            {
                collection: collection,
                internal_id: internalId,
                external_id: externalId
            }, (err, res) => {
                callback(err);
            }
        );
    }


}

src/version1/MappingsNullClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, DataPage } from "pip-services3-commons-node";
import { MappingV1 } from "./MappingV1";


export class MappingsNullClientV1 implements IMappingsClient {
    getCollectionNames(correlationId: string, callback: (err: any, items: string[]) => void) {
        callback(null, new Array<string>());
    }
    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams, callback: (err: any, page: DataPage<MappingV1>) => void) {
        callback(null, new DataPage<MappingV1>());
    }
    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number, callback: (err: any) => void) {
        callback(null);
    }
    mapToExternal(correlationId: string, collection: string, internalId: string, callback: (err: any, externalId: string) => void) {
        callback(null, null);
    }
    mapToInternal(correlationId: string, collection: string, externalId: string, callback: (err: any, internalId: string) => void) {
        callback(null, null);
    }
    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string, callback: (err: any) => void) {
        callback(null);
    }

}

src/version1/MappingsDirectClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, Descriptor } from "pip-services3-commons-node";
import { DirectClient } from 'pip-services3-rpc-node';

export class MappingsDirectClientV1 extends DirectClient<any> implements IMappingsClient {
    
    public constructor() {
        super();
        this._dependencyResolver.put('controller', new Descriptor('pip-services-mappings', 'controller', '*', '*', '1.0'));
    }

    getCollectionNames(correlationId: string, callback: (err: any, items: string[]) => void) {
        let timing = this.instrument(correlationId, 'mappings.get_collections_names');
        this._controller.getCollectionNames(correlationId, (err, items) => {
            timing.endTiming();
            callback(err, items);
        });
    }

    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams, callback: (err: any, page: import("pip-services3-commons-node").DataPage<import("./MappingV1").MappingV1>) => void) {
        let timing = this.instrument(correlationId, 'mappings.get_mappings');
        this._controller.getMappings(correlationId, filter, paging, (err, page) => {
            timing.endTiming();
            callback(err, page);
        });
    }

    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number, callback: (err: any) => void) {
        let timing = this.instrument(correlationId, 'mappings.add_mapping');
        this._controller.addMapping(correlationId, collection, internalId, externalId, timeToLive, (err) => {
            timing.endTiming();
            callback(err);
        });
    }

    mapToExternal(correlationId: string, collection: string, internalId: string, callback: (err: any, externalId: string) => void) {
        let timing = this.instrument(correlationId, 'mappings.map_to_external');
        this._controller.mapToExternal(correlationId, collection, internalId, (err, id) => {
            timing.endTiming();
            callback(err, id);
        });
    }

    mapToInternal(correlationId: string, collection: string, externalId: string, callback: (err: any, internalId: string) => void) {
        let timing = this.instrument(correlationId, 'mappings.map_to_internal');
        this._controller.mapToInternal(correlationId, collection, externalId, (err, id) => {
            timing.endTiming();
            callback(err, id);
        });
    }

    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string, callback: (err: any) => void) {
        let timing = this.instrument(correlationId, 'mappings.delete_mapping');
        this._controller.deleteMapping(correlationId, collection, internalId, externalId, (err) => {
            timing.endTiming();
            callback(err);
        });
    }

}

src/version1/MappingsMemoryClientV1.ts
let _ = require('lodash');

import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, DataPage, IdGenerator } from "pip-services3-commons-node";
import { MappingV1 } from "./MappingV1";

export class MappingsMemoryClientV1 implements IMappingsClient {
    private _maxPageSize: number = 100;
    private _items: MappingV1[];
    private readonly _defaultTTL: number = 7 * 24 * 60 * 60 * 1000;

    public constructor(...items: MappingV1[]) {
        this._items = items;
    }

    private composeFilter(filter: FilterParams): any {
        filter = filter || new FilterParams();
        let collection = filter.getAsNullableString("collection");
        let id = filter.getAsNullableString("id");
        let internalId = filter.getAsNullableString("internal_id");
        let externalId = filter.getAsNullableString("external_id");
        let search = filter.getAsNullableString("search");

        return (item: MappingV1) => {
            if (collection != null && item.collection != collection)
                return false;
            if (id != null && item.external_id != id && item.internal_id != id)
                return false;
            if (internalId != null && item.internal_id != internalId)
                return false;
            if (externalId != null && item.external_id != externalId)
                return false;
            if (search != null && item.external_id != search && item.internal_id != search)
                return false;
            return true;
        };
    }

    getCollectionNames(correlationId: string, callback: (err: any, items: string[]) => void) {
        let result: string[] = [];
        for (let mapping of this._items) {
            let collection = mapping.collection;
            if (result.indexOf(collection) < 0)
                result.push(collection);
        }
        callback(null, result);
    }

    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams, callback: (err: any, page: import("pip-services3-commons-node").DataPage<import("./MappingV1").MappingV1>) => void) {
        let filterMappings = this.composeFilter(filter);
        let mappings = _.filter(this._items, filterMappings);

        // Extract a page
        paging = paging != null ? paging : new PagingParams();
        let skip = paging.getSkip(-1);
        let take = paging.getTake(this._maxPageSize);

        let total = null;
        if (paging.total)
            total = mappings.length;

        if (skip > 0)
            mappings = _.slice(mappings, skip);
        mappings = _.take(mappings, take);

        let page = new DataPage<MappingV1>(mappings, total);
        callback(null, page);
    }

    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number, callback: (err: any) => void) {
        let mapping: MappingV1;
        timeToLive = timeToLive > 0 ? timeToLive : this._defaultTTL;
        mapping = <MappingV1>{
            collection: collection,
            internal_id: internalId,
            external_id: externalId,
            expiration_time: new Date(new Date().getTime() + timeToLive)
        };
        if (mapping == null) {
            callback(null);
            return;
        }
        mapping = _.clone(mapping);
        mapping.id = mapping.id || IdGenerator.nextLong();
        this._items.push(mapping);
        callback(null);
    }

    mapToExternal(correlationId: string, collection: string, internalId: string, callback: (err: any, externalId: string) => void) {

        let result: string = null;

        let items = _.filter(this._items, (m) => {
            return collection.localeCompare(m.collection) == 0 && internalId.localeCompare(m.internal_id) == 0
        });
        let mapping: MappingV1 = items.length > 0 ? items[0] : null;
        result = mapping != null && mapping.expiration_time > new Date() ? mapping.external_id : null;
        callback(null, result);
    }

    mapToInternal(correlationId: string, collection: string, externalId: string, callback: (err: any, internalId: string) => void) {

        let result: string = null;
        let items = _.filter(this._items, (m) => {
            return collection.localeCompare(m.collection) == 0 && externalId.localeCompare(m.external_id) == 0
        });
        let mapping: MappingV1 = items.length > 0 ? items[0] : null;
        result = mapping != null && mapping.expiration_time > new Date() ? mapping.internal_id : null;
        callback(null, result);
    }

    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string, callback: (err: any) => void) {

        for (let index = this._items.length - 1; index >= 0; index--) {
            let mapping = this._items[index];
            if (mapping.collection == collection
                && mapping.internal_id == internalId
                && mapping.external_id == externalId) {
                this._items.splice(index, 1);
                break;
            }
        }

        callback(null);
    }


}